### ISSUE TITLE
incorrect garbage collection on initialization to self through    reference

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4650]
Migrated from [rt.perl.org#8001](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8001) (status was 'resolved')

Searchable as RT8001$

**Comment by p5pRT** [comment]
### From ksibson@matrix.net

#### Created by ksibson@matrix.net



A variable being modified in a subroutine is initialized to itself through a reference on the
subroutine's return\. In principle this should be legal\, though it's a somewhat unusual thing to want to do\. 

Here's an equivalent but minimal example&#8203;:

\-\-
my %hash = %\{doit\(\)\};

sub doit \{
 $hash\{'foo'\} = 'bar';
 return \\%hash;
\}

print "KEYS&#8203;: "\, keys %hash\, "\\n";
print "DEF : "\, \(defined $hash\{'foo'\}\)?'defined'&#8203;:'undefined'\, "\\n";
print "VALS&#8203;: "\, values %hash\, "\\n";
\-\-

On the initialization \(basically to itself\) the Perl interpreter
blows away whatever memory the original had associated with it\, 
believing I suppose that it's no longer referenced\.

In this simple code $hash\{'foo'\} becomes undefined\, but in larger
programs it crashes unpredictably\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.1:

Configured by rah at Thu Oct  4 21:31:59 CDT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=solaris, osvers=2.8, archname=sun4-solaris
    uname='sunos green 5.8 generic_108528-07 sun4u sparc sunw,ultra-5_10 '
    config_args='-des -Dusemymalloc -Dprefix=/usr/local/perl'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.2 19991024 (release)', gccosandvers='solaris2.8'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib '
    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/perl/lib/5.6.1/sun4-solaris
    /usr/local/perl/lib/5.6.1
    /usr/local/perl/lib/site_perl/5.6.1/sun4-solaris
    /usr/local/perl/lib/site_perl/5.6.1
    /usr/local/perl/lib/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/ksibson
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/proc/bin:/bin:/opt/mysql/bin:/usr/java1.2/bin:/home/ksibson/bin/:/usr/local/perl/bin:/usr/local/bin:/usr/local/ssl/bin:/usr/ccs/bin:/usr/j2se/bin:/usr/java/bin:/opt/tomcat/bin:/usr/bin:/usr/proc/bin:/usr/xpg4/bin:/usr/ucb
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @floatingatoll

Summary&#8203;: this bug was introduced between 5\.5\.3 and 5\.6\.0\, and exists on
all platforms I tested it on\.

R\.

Running this test program&#8203;:

\-\-\-\- begin program

my %hash = %\{doit\(\)\};

sub doit \{
 $hash\{'foo'\} = 'bar';
 return \\%hash;
\}

print "KEYS&#8203;: "\, keys %hash\, "\\n";
print "DEF : "\, \(defined $hash\{'foo'\}\)?'defined'&#8203;:'undefined'\, "\\n";
print "VALS&#8203;: "\, values %hash\, "\\n";

\-\-\-\- end program

returns the following results on these versions of perl&#8203;:

5\.005\_03 \(redhat 6\.2\, FreeBSD 4\.1\-RELEASE\)&#8203;:

KEYS&#8203;: foo
DEF : defined
VALS&#8203;: bar

5\.6\.0 \(OpenBSD 2\.9\, 2 registered patches \-
&nbsp;        SUIDMAIL \- fixes for suidperl security
&nbsp;        CHANGE6214 \- fixes a coredump in @&#8203;DB&#8203;::args\)\,
5\.6\.1 \(activestate build 626 on windows\, pure 5\.6\.1 on linux\-2\.4\.7\)\,
perl\-current \(@&#8203;13529\) and perl\-stable \(@&#8203;13478\)&#8203;:

KEYS&#8203;: foo
DEF : undefined
VALS&#8203;:

Looks like it was introduced between 5\.005\_03 and 5\.6\.0\.

This is a memory corruption bug\, not a scoping bug\, right?

R\.

**Comment by p5pRT** [comment]
### From @rgarcia

On 2001\.12\.07 01&#8203;:20 Keith Sibson wrote&#8203;:
> 
> A variable being modified in a subroutine is initialized to itself through a reference on the
> subroutine's return\. In principle this should be legal\, though it's a somewhat unusual thing to want to do\. 
> 
> Here's an equivalent but minimal example&#8203;:
> 
> \-\-
> my %hash = %\{doit\(\)\};
> 
> sub doit \{
>  $hash\{'foo'\} = 'bar';
>  return \\%hash;
> \}
> 
> print "KEYS&#8203;: "\, keys %hash\, "\\n";
> print "DEF : "\, \(defined $hash\{'foo'\}\)?'defined'&#8203;:'undefined'\, "\\n";
> print "VALS&#8203;: "\, values %hash\, "\\n";
> \-\-
> 
> On the initialization \(basically to itself\) the Perl interpreter
> blows away whatever memory the original had associated with it\, 
> believing I suppose that it's no longer referenced\.

Not really\, in my opinion\. The hash remains\, its data goes away\.
\(You will notice that the 'foo' key still exists in %hash\)\.

This is due to the dual nature of my\(\)\.
It has the compile\-time effect to declare a lexical scope for %hash\,
and the run\-time effect to initialize it\. The doit\(\) subroutine
being in the scope where my %hash is valid\, "use strict 'vars'"
doesn't complain\. But it's called at a moment where the run\-time
effect of my\(\) has not occured yet\.

perlsub says :

&nbsp;  \.\.\.
&nbsp;  The declared variable is not introduced \(is not visible\) until after
&nbsp;  the current statement\.  Thus\,

&nbsp;      my $x = $x;

&nbsp;  can be used to initialize a new $x with the value of the old $x\, and
&nbsp;  the expression

&nbsp;      my $x = 123 and $x == 123

&nbsp;  is false unless the old $x happened to have the value C\<123>\.
&nbsp;  \.\.\.

Thus the doit\(\) uses the pad to store the autovivified %hash\. Then\,
the pad entry is overwritten by a copy of itself \(in the
%hash=%\{doit\(\)\} assignment\)\. Hence the %hash doesn't
change\, but the data stored into the hash goes away\. The my\(\) run\-time
effect has not occured yet and perl considers that this data is
no longer referenced\. If I understand correctly\.

> In this simple code $hash\{'foo'\} becomes undefined\, but in larger
> programs it crashes unpredictably\.

More annoying indeed\.
